packages/node-core/src/indexer/multiChainRewind.service.ts

```diff
@@ -0,0 +1,367 @@
+            } satisfies MultiChainRewindPayload);
+            break;
+          }
+          case MultiChainRewindEvent.RewindComplete:
+            // recover indexing status
+            this.status = RewindStatus.Normal;
+            this.waitRewindHeader = undefined;
+            break;
+          default:
+            throw new Error(`Unknown rewind event: ${eventType}`);
+        }
+        logger.info(`Handle success rewind event: ${eventType}, chainId: ${this.chainId}`);
+      });
+    });
+
+    await this.pgListener.query(`LISTEN "${this.rewindTriggerName}"`);
+    logger.info(`Register rewind listener success, chainId: ${this.chainId}`);
+
+    // Check whether the current state is in rollback.
+    const {rewindLock, rewindTimestamp} = await this.getGlobalRewindStatus();
+    if (rewindLock) {
+      this.status = RewindStatus.WaitOtherChain;
+    }
+    if (rewindTimestamp) {
+      this.status = RewindStatus.Rewinding;
+      this.waitRewindHeader = await this.getHeaderByBinarySearch(dayjs(rewindTimestamp).toDate());
+    }
+  }
+
+  /**
+   * Serialize the rewind lock
+   * @param rewindTimestamp ms
+   * @param chainTotal The total number of registered chains.
+   * @returns
+   */
+  private serializeRewindLock(rewindTimestamp: number, chainTotal: number): string {
+    return JSON.stringify({timestamp: rewindTimestamp, chainNum: chainTotal});
+  }
+
+  async getGlobalRewindStatus() {
+    const rewindTimestampKey = generateRewindTimestampKey(this.chainId);
+
+    const records = await this.storeService.globalDataRepo.findAll({
+      where: {key: {[Op.in]: [rewindTimestampKey, RewindLockKey]}},
+    });
+    const rewindLockInfo: GlobalData<typeof RewindLockKey> | undefined = records
+      .find((r) => r.key === RewindLockKey)
+      ?.toJSON();
+    const rewindTimestampInfo: GlobalData<RewindTimestampKey> | undefined = records
+      .find((r) => r.key === rewindTimestampKey)
+      ?.toJSON();
+
+    assert(
+      rewindTimestampInfo !== undefined,
+      `Not registered rewind timestamp key in global data, chainId: ${this.chainId}`
+    );
+    return {rewindTimestamp: rewindTimestampInfo.value, rewindLock: rewindLockInfo?.value};
+  }
+
+  /**
+   * If the set rewindTimestamp is greater than or equal to the current blockHeight, we do nothing because we will roll back to an earlier time.
+   * If the set rewindTimestamp is less than the current blockHeight, we should roll back to the earlier rewindTimestamp.
+   * @param rewindTimestamp rewindTimestamp in milliseconds
+   */
+  async setGlobalRewindLock(rewindTimestamp: number) {
+    const globalTable = this.storeService.globalDataRepo.tableName;
+    const chainTotal = await this.storeService.globalDataRepo.count({
+      where: {
+        key: {[Op.like]: `${RewindTimestampKeyPrefix}_%`},
+      },
+    });
+
+    const tx = await this.sequelize.transaction();
+    try {
+      const [_, updateRows] = await this.sequelize.query(
```

Comment:
I think these global table db operations can be moved into a PlainGlobalModel. It would be good to have all db operations in a single location

Reply:
Could this query string be generalised between metadata?